python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,432 @@
+              Otherwise, (None, None, None)
+    """
+    try:
+        from .ui.qt_abstraction import QtGui, QtCore
+    except ImportError:
+        return None, None, None
+    return QtCore, QtGui, QtGui.QApplication.instance() is not None
+
+
+def _create_invoker():
+    """
+    Create the object used to invoke function calls on the main thread when
+    called from a different thread.
+
+    :returns:  Invoker instance. If Qt is not available or there is no UI, no invoker will be returned.
+    """
+    QtCore, QtGui, has_ui = _get_qt_state()
+    # If we have a ui and we're not in the main thread, we'll need to send ui requests to the
+    # main thread.
+    if not QtCore or not QtGui or not has_ui:
+        return lambda fn, *args, **kwargs: fn(*args, **kwargs)
+
+    # If we are already in the main thread, no need for an invoker, invoke directly in this thread.
+    if QtCore.QThread.currentThread() == QtGui.QApplication.instance().thread():
+        return lambda f: f()
+
+    class MainThreadInvoker(QtCore.QObject):
+        """
+        Class that allows sending message to the main thread.
+        """
+        def __init__(self):
+            """
+            Constructor.
+            """
+            QtCore.QObject.__init__(self)
+            self._res = None
+            self._exception = None
+            # Make sure that the invoker is bound to the main thread
+            self.moveToThread(QtGui.QApplication.instance().thread())
+
+        def __call__(self, fn, *args, **kwargs):
+            """
+            Asks the MainTheadInvoker to call a function with the provided parameters in the main
+            thread.
+            :param fn: Function to call in the main thread.
+            :param args: Array of arguments for the method.
+            :param kwargs: Dictionary of named arguments for the method.
+            :returns: The result from the function.
+            """
+            self._fn = lambda: fn(*args, **kwargs)
+            self._res = None
+
+            logger.debug("Sending ui request to main thread.")
+
+            QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
+
+            # If an exception has been thrown, rethrow it.
+            if self._exception:
+                raise self._exception
+            return self._res
+
+        @QtCore.Slot()
+        def _do_invoke(self):
+            """
+            Execute function and return result
+            """
+            try:
+                logger.debug("Invoking from main thread.")
+                self._res = self._fn()
+            except Exception, e:
+                self._exception = e
+
+    return MainThreadInvoker()
+
+
+class AuthenticationHandlerBase(object):
+    """
+    Base class for authentication requests. It handles locking reading cached credentials
+    on disk and writing newer credentials back. It also keeps track of any attempt to cancel
```

Comment:
These classes (and all their derived friends) should only be created by those factory methods down the bottom right? May be worth pointing that out in the docstrings.